server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java

```diff
@@ -0,0 +1,291 @@
+  }
+
+  private static <OUTPUT, INPUT extends Comparable<INPUT>, RESULT extends Collection<OUTPUT>> RESULT executeLargeInputs(Collection<INPUT> input,
+    Function<List<INPUT>, RESULT> function, java.util.function.Function<Integer, RESULT> outputInitializer) {
+    if (input.isEmpty()) {
+      return outputInitializer.apply(0);
+    }
+    RESULT results = outputInitializer.apply(input.size());
+    for (List<INPUT> partition : toUniqueAndSortedPartitions(input)) {
+      RESULT subResults = function.apply(partition);
+      if (subResults != null) {
+        results.addAll(subResults);
+      }
+    }
+    return results;
+  }
+
+  /**
+   * Partition by 1000 elements a list of input and execute a consumer on each part.
+   *
+   * The goal is to prevent issue with ORACLE when there's more than 1000 elements in a 'in ('X', 'Y', ...)'
+   * and with MsSQL when there's more than 2000 parameters in a query
+   */
+  public static <INPUT extends Comparable<INPUT>> void executeLargeUpdates(Collection<INPUT> inputs, Consumer<List<INPUT>> consumer) {
+    Iterable<List<INPUT>> partitions = toUniqueAndSortedPartitions(inputs);
+    for (List<INPUT> partition : partitions) {
+      consumer.accept(partition);
+    }
+  }
+
+  /**
+   * Partition by 1000 elements a list of input and execute a consumer on each part.
+   *
+   * The goal is to prevent issue with ORACLE when there's more than 1000 elements in a 'in ('X', 'Y', ...)'
+   * and with MsSQL when there's more than 2000 parameters in a query
+   *
+   * @param sqlCaller a {@link Function} which calls the SQL update/delete and returns the number of updated/deleted rows.
+   *
+   * @return the total number of updated/deleted rows (computed as the sum of the values returned by {@code sqlCaller}).
+   */
+  public static <INPUT extends Comparable<INPUT>> int executeLargeUpdates(Collection<INPUT> inputs, Function<List<INPUT>, Integer> sqlCaller) {
+    Iterable<List<INPUT>> partitions = toUniqueAndSortedPartitions(inputs);
+    Integer res = 0;
+    for (List<INPUT> partition : partitions) {
+      res += sqlCaller.apply(partition);
+    }
+    return res;
+  }
+
+  /**
+   * Ensure values {@code inputs} are unique (which avoids useless arguments) and sorted before creating the partition.
+   */
+  private static <INPUT extends Comparable<INPUT>> Iterable<List<INPUT>> toUniqueAndSortedPartitions(Collection<INPUT> inputs) {
+    return Iterables.partition(toUniqueAndSortedList(inputs), PARTITION_SIZE_FOR_ORACLE);
+  }
+
+  /**
+   * Ensure values {@code inputs} are unique (which avoids useless arguments) and sorted so that there is little
+   * variations of SQL requests over time as possible with a IN clause and/or a group of OR clauses. Such requests can
+   * then be more easily optimized by the SGDB engine.
+   */
+  public static <INPUT extends Comparable<INPUT>> List<INPUT> toUniqueAndSortedList(Iterable<INPUT> inputs) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)